### ISSUE TITLE
MD043: Add support for grouping for CHANGELOG scenario (repeated heading sets)

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [DavidAnson/markdownlint#1601]
[MD024 / no-duplicate-heading](https://github.com/DavidAnson/markdownlint/blob/224987d727ebada36752074d2466a061c9e03ff0/doc/Rules.md#md024---multiple-headings-with-the-same-content) includes a `siblings_only` property:

https://github.com/DavidAnson/markdownlint/blob/224987d727ebada36752074d2466a061c9e03ff0/doc/Rules.md?plain=1#L806

Is there a reason this is on by default? Duplicate _non-sibling_ headings are not necessarily an issue. For common terms like `## API` -> `### (feature)` -> `#### Examples`, it can sometimes make more sense to be consistent with headings across sections.

Example resource discussing how non-sibling duplicates are not necessarily an accessibility issue: https://www.tpgi.com/heading-off-confusion-when-do-headings-fail-wcag

Related issues: #300, #1591

Related comment in the wild: https://github.com/JoshuaKGoldberg/package-json-validator/pull/230#discussion_r2085067204

**Comment by DavidAnson** [comment]
That setting did not used to exist and was added in response to the changelog pattern you cite. Because the rule is ultimately about not duplicating headings, the default behavior is to not duplicate ANY headings. For people in the nested scenario, they can opt in.

**Comment by JoshuaKGoldberg** [comment]
> the default behavior is to not duplicate ANY headings

Can I ask, why? 🙂

As in, what is the set of reasons to enable the rule - and which of those reasons still apply to non-siblings?

**Comment by DavidAnson** [comment]
Quoting the documentation:

> Rationale: Some Markdown parsers generate anchors for headings based on the heading name; headings with the same content can cause problems with that.

I can't say which parsers, or whether they still have this problem.

**Comment by maettu-this** [comment]
> Some Markdown parsers generate anchors for headings based on the heading name

Proposing to question whether this indeed is still the case, or just some legacy issue. I am having the very same issue:

Technical documents tend to repeat a structure, e.g.:

### Functional Requirements

#### Scope A
- Requirement A1
- Requirement A2
- ...
#### Scope B
- Requirement B1
- Requirement B2
- ...

### Non-Functional Requirements

#### Scope A
- Requirement A1
- Requirement A2
- ...
#### Scope B
- Requirement B1
- Requirement B2
- ...

I'd also prefer `siblings_only` be enabled by default.

**Comment by DavidAnson** [comment]
Looking at this issue a bit more, I don't think defaulting `siblings_only` to `true` is the right change. Yes, it helps with the CHANGELOG scenario, but doing that allows TOO much. For example, this tweak to the documentation sample would no longer produce violations by default:

```markdown
# Change log

## 1.0.0

### Features

## 2.0.0

### Features

## Features

^^ WRONG: LEVEL

### Change log

^^ WRONG: DUPLICATED

<!-- markdownlint-configure-file {
  "no-duplicate-heading": {
    "siblings_only": true
  }
} -->
```

For a rule called `no-duplicate-heading`, the above behavior seems undesirable.

So far, the only compelling scenario I've seen to change the current behavior of this rule is the CHANGELOG scenario - but I think a better way to address that might be a version of `MD043`/`required-headings` that has been enhanced to allow repeated GROUPS of headings (similar to existing support for `*`/`+`/`?`, but more complicated).

<https://github.com/DavidAnson/markdownlint/blob/main/doc/md043.md>

**Comment by maettu-this** [comment]
David, thanks for looking into this once more. Given...

> https://github.com/DavidAnson/markdownlint/blob/main/doc/md043.md

...I don't yet understand how I could configure the [earlier commented](https://github.com/DavidAnson/markdownlint/issues/1601#issuecomment-4528571152) example to prevent MD024 from being issued.

**Comment by DavidAnson** [comment]
If using a possible future version of MD043 that supported grouping for a change log scenario, I expect you would disable MD024.
